README.md
# vue-adaptive

A rem scale adaptive with breakpoints for Vue.js

[![npm](https://img.shields.io/npm/v/vue-adaptive/latest)](https://www.npmjs.com/package/vue-adaptive)
[![npm](https://img.shields.io/npm/dt/vue-adaptive)](https://www.npmjs.com/package/vue-adaptive)
[![npm bundle size (version)](https://img.shields.io/bundlephobia/min/vue-adaptive/latest)](https://bundlephobia.com/result?p=vue-adaptive@latest)
[![Inline docs](http://inch-ci.org/github/zmey3301/vue-adaptive.svg?branch=v2)](http://inch-ci.org/github/zmey3301/vue-adaptive?branch=v2)
[![MIT](https://img.shields.io/npm/l/vue-adaptive.svg)](https://opensource.org/licenses/MIT)
[![GitHub issues](https://img.shields.io/github/issues/zmey3301/vue-adaptive.svg)](https://github.com/zmey3301/vue-adaptive/issues)
[![GitHub last commit](https://img.shields.io/github/last-commit/zmey3301/vue-adaptive/v2)](https://github.com/zmey3301/vue-adaptive/commits/master)
[![David](https://img.shields.io/david/zmey3301/vue-adaptive)](https://david-dm.org/zmey3301/vue-adaptive)
[![CodeFactor](https://www.codefactor.io/repository/github/zmey3301/vue-adaptive/badge)](https://www.codefactor.io/repository/github/zmey3301/vue-adaptive)
[![codecov](https://codecov.io/gh/zmey3301/vue-adaptive/branch/master/graph/badge.svg?token=rDPIiIiKyl)](https://codecov.io/gh/zmey3301/vue-adaptive)
[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)

## Setup

Install vue-adaptive from **npm**: `npm i vue-adaptive --save`

Or **yarn**: `yarn add vue-adaptive`

Then import plugin and use in your main.js

```javascript
import Vue from 'vue'
import Adaptive from 'vue-adaptive'

Vue.use(Adaptive, config)
```

## Configuration

Configuration object has two parts:

#### 1. Global configuration

Configuration object has object named 'global' in it. This object contains all global parameters of plugin.

```json
{
  "global": {
    "rem": 10,
    "throttling": 17,
    "orientationTestCount": 25,
    "orientationChangeTimeout": 1000
  }
}
```

`rem`: Default rem value in case no breakpoint matched. Default: 10.

`throttling`: Time in milliseconds, changes the frequency of viewport update.
May be useful if you don't care about changes step and want to reduce CPU usage. Default: 17 (60fps frame time).

`orientationTestCount`: Count of checks with identical results meaning orientationchange event is over. Default: 25.

`orientationChangeTimeout`: Max duration of the orientationchange event. Default: 1000.

#### 2. Devices (breakpoints) configuration

Besides 'global' object configuration object represents a list of breakpoints (or devices)

Device structure is:

```
<device>: {// Device (and global class) name
    if: <if>,
	element: <selector or HTMLElement>,
	rem: <px>
	from: {
		width: <fromWidth>,
		height: <fromHeight>
	},
	to: {
		width: <toWidth>,
		height: <toHeight>
	},
	base: {
		width: <baseWidth>,
		height: <baseHeight>
	},
	k: <k>
}
```

`element`: String or HTMLElement, contains element selector that will be used for detecting viewport. Default: window.

`if`: Condition, `functions(viewport)`, viewport contains width and height of element, function affects device type.

`rem`: Static rem value, overrides `k` and `base`.

`from`: Minimal width and height for setting device.

`to`: Maximal width and height for setting device.

`base`: Base width and height for rem calculation.

`k`: Additional k for rem calculation. Default: 1.

`setDevice`: Boolean, if false adaptive will not set `device` or `no-device` class on the html element,
and deviceType based on device name. Default: true.

You also can create many breakpoints for one device by adding name to device, for example:

```json
{
  "desktop:wide": {
    "rem": 10,
    "from": {
      "width": 1366
    }
  },
  "desktop:thin": {
    "k": 0.75,
    "from": {
      "width": 1008
    },
    "base": {
      "width": 1100
    },
    "to": {
      "width": 1365
    }
  }
}
```

In this case both breakpoints will have `desktop` global class (and `$adaptive.is` of course), but two various rem calculations.

## Usage

You can use global classes in your styles:

```stylus
.someblock
    background-color: blue
    .mobile &
        background-color: red
```

Or conditions in vue template:

```vue
<div v-show="$adaptive.is.mobile"></div>
```

You always can see current viewport and rem value in `$adaptive` inside of your component.

```js
computed = {
	isWidth1600 () {
		return this.$adaptive.width >= 1600
	}
}
```

`$adaptive` has `height`, `width`, `rem` numeric values and `is` object, containing all device names with boolean value.


CHANGELOG.md
## [2.0.2](https://github.com/zmey3301/vue-adaptive/compare/v2.0.1...v2.0.2) (2021-01-31)


### Build System

* **npm bundle:** updated npmignore and readme badges ([4951c61](https://github.com/zmey3301/vue-adaptive/commit/4951c61))


### Continuous Integration

* **github releases:** fixed binaries installation ([87b381e](https://github.com/zmey3301/vue-adaptive/commit/87b381e))



## [2.0.1](https://github.com/zmey3301/vue-adaptive/compare/v2.0.0...v2.0.1) (2021-01-30)


### Continuous Integration

* **github releases:** added GitHub release automation ([68c9192](https://github.com/zmey3301/vue-adaptive/commit/68c9192))
* **github releases:** fixed CI configuration ([a3489de](https://github.com/zmey3301/vue-adaptive/commit/a3489de))



# [2.0.0](https://github.com/zmey3301/vue-adaptive/compare/v2.0.0-1...v2.0.0) (2021-01-30)


### Bug Fixes

* **changelog prettier ignore:** prettier configuration update for autogenerated changelog ignore ([4c765bf](https://github.com/zmey3301/vue-adaptive/commit/4c765bf))
* **npm version hook:** version release process fix ([a51ae48](https://github.com/zmey3301/vue-adaptive/commit/a51ae48))


### Chores

* **npm hooks:** added NPM hook for changelog generation ([4c2bc03](https://github.com/zmey3301/vue-adaptive/commit/4c2bc03))
* **project conf:** rEADME/CHANGELOG updates and commitizen init ([8407961](https://github.com/zmey3301/vue-adaptive/commit/8407961))


### Continuous Integration

* **circleci config:** circleCI runners node version bump ([9818a81](https://github.com/zmey3301/vue-adaptive/commit/9818a81))


### Styles

* **circleci config:** circleCI config auto-format ([d993fe2](https://github.com/zmey3301/vue-adaptive/commit/d993fe2))








tests/adaptive.test.js
import {createLocalVue, mount} from "@vue/test-utils"

import Adaptive from "../src"
import config from "./config.json"
import Test from "./test.vue"

function resizeViewport(width, height) {
  if (width) window.innerWidth = width
  if (height) window.innerHeight = height
  window.dispatchEvent(new Event("resize"))
}

function rotateDevice(duration = 500, steps = 100) {
  const widthStep = (window.innerHeight - window.innerWidth) / steps
  const heightStep = (window.innerWidth - window.innerHeight) / steps
  function orientationChange() {
    window.innerWidth += widthStep
    window.innerHeight += heightStep
  }
  const interval = setInterval(() => {
    steps--
    if (steps > 0) orientationChange()
    else clearInterval(interval)
  }, duration / steps)
  orientationChange()
  window.dispatchEvent(new Event("orientationchange"))
}

const resizeHandleTimeout = 50
const localVue = createLocalVue()
localVue.use(Adaptive, config)

describe("Adaptive events handling", () => {
  const wrapper = mount(Test, {localVue})

  beforeEach(() => {
    resizeViewport(1024, 768)
  }, resizeHandleTimeout)

  test("Vue.adaptive should be equal to the instances $adaptive", () => {
    expect(localVue.adaptive).toEqual(wrapper.vm.$adaptive)
  })

  test("Vue.adaptive should stay equal to the instances $adaptive after resize", done => {
    const width = 500
    resizeViewport(width)
    setTimeout(() => {
      try {
        const cond = expect(wrapper.vm.$adaptive.width)
        cond.toBe(localVue.adaptive.width)
        cond.toBe(width)
        done()
      } catch (error) {
        done(error)
      }
    }, resizeHandleTimeout)
  })

  test("Adaptive should react to resize event", done => {
    const size = 700
    resizeViewport(size, size)
    setTimeout(() => {
      try {
        expect(localVue.adaptive.width).toBe(size)
        expect(localVue.adaptive.height).toBe(size)
        done()
      } catch (error) {
        done(error)
      }
    }, resizeHandleTimeout)
  })

  test("Adaptive should process orientationchange event", done => {
    const initialWidth = localVue.adaptive.width
    const initialHeight = localVue.adaptive.height
    const orientationChangeDuration = 500
    rotateDevice(orientationChangeDuration)
    setTimeout(() => {
      try {
        const initialDimensions = [initialWidth, initialHeight]
        expect(localVue.adaptive.width).not.toBeOneOf(initialDimensions)
        expect(localVue.adaptive.height).not.toBeOneOf(initialDimensions)
        setTimeout(() => {
          try {
            expect(localVue.adaptive.width).toBeCloseTo(initialHeight)
            expect(localVue.adaptive.height).toBeCloseTo(initialWidth)
            done()
          } catch (error) {
            done(error)
          }
        }, orientationChangeDuration * 1.5)
      } catch (error) {
        done(error)
      }
    }, (1000 / 60) * 1.5)
  })

  test("Adaptive should stop orientation change handle on timeout", done => {
    const initialWidth = localVue.adaptive.width
    const initialHeight = localVue.adaptive.height
    const orientationChangeDuration = 1500

    rotateDevice(orientationChangeDuration)
    setTimeout(() => {
      try {
        expect(localVue.adaptive.width).not.toBeCloseTo(initialHeight)
        expect(localVue.adaptive.height).not.toBeCloseTo(initialWidth)
        done()
      } catch (error) {
        done(error)
      }
    }, 2000)
  })
})

describe("Adaptive breakpoint matching", () => {
  test("Adaptive should select breakpoint on dimensions match", done => {
    resizeViewport(1000, 600)
    setTimeout(() => {
      try {
        expect(localVue.adaptive.is).toEqual({
          desktop: false,
          tablet: true,
          mobile: false
        })

        expect(document.documentElement.className).toInclude("tablet")
        expect(document.documentElement.className).toIncludeMultiple([
          "no-desktop",
          "no-mobile"
        ])

        expect(document.documentElement.className).not.toInclude("no-tablet")
        done()
      } catch (error) {
        done(error)
      }
    }, resizeHandleTimeout)
  })

  test("Adaptive should not select breakpoint on one dimension match", done => {
    function checkBreakpoints() {
      expect(localVue.adaptive.is.tablet).toBeFalse()
      expect(document.documentElement.className).toInclude("no-tablet")
    }
    resizeViewport(1000, 1000)
    setTimeout(() => {
      try {
        checkBreakpoints()
        resizeViewport(1100, 600)
        setTimeout(() => {
          try {
            checkBreakpoints()
            done()
          } catch (error) {
            done(error)
          }
        }, resizeHandleTimeout)
      } catch (error) {
        done(error)
      }
    }, resizeHandleTimeout)
  })

  test("Adaptive should select breakpoint on exclusive lower match", done => {
    resizeViewport(1008)
    setTimeout(() => {
      try {
        expect(localVue.adaptive.is.desktop).toBeTrue()
        done()
      } catch (error) {
        done(error)
      }
    }, resizeHandleTimeout)
  })

  test("Adaptive should select breakpoint on inclusive higher match", done => {
    resizeViewport(1440)
    setTimeout(() => {
      try {
        expect(localVue.adaptive.is.desktop).toBeFalse()
        done()
      } catch (error) {
        done(error)
      }
    }, resizeHandleTimeout)
  })

  test("Adaptive should select breakpoint if only one of suffixed breakpoints match", done => {
    resizeViewport(1500, 768)
    setTimeout(() => {
      try {
        expect(localVue.adaptive.is.desktop).toBeTrue()
        expect(document.documentElement.className).toInclude("desktop")
        expect(document.documentElement.className).not.toInclude("no-desktop")
        done()
      } catch (error) {
        done(error)
      }
    }, resizeHandleTimeout)
  })

  test("Adaptive should not add device classes if noDevice set to false", () => {
    expect(localVue.adaptive.is).not.toHaveProperty("body")
    expect(document.documentElement.className).not.toMatch(
      /(?:^| )(?:no-)?body(?:$| )/
    )
  })
})

describe("Adaptive rem calculation", () => {
  test("Adaptive should set default rem in case no breakpoints matched", done => {
    resizeViewport(50, 50)
    setTimeout(() => {
      try {
        expect(localVue.adaptive.rem).toBe(12.5)
        expect(document.documentElement.style.fontSize).toBe("12.5px")
        done()
      } catch (error) {
        done(error)
      }
    }, resizeHandleTimeout)
  })

  test("Adaptive should set static rem based on config", done => {
    resizeViewport(1500)
    setTimeout(() => {
      try {
        expect(localVue.adaptive.rem).toBe(10)
        expect(document.documentElement.style.fontSize).toBe("10px")
        done()
      } catch (error) {
        done(error)
      }
    }, resizeHandleTimeout)
  })

  test("Adaptive should calculate rem based on base object", done => {
    const width = 1200
    resizeViewport(width)
    setTimeout(() => {
      try {
        const rem = (width / 1008) * 10 * 0.75 // device k
        expect(localVue.adaptive.rem).toBeCloseTo(rem)
        expect(parseFloat(document.documentElement.style.fontSize)).toBeCloseTo(
          rem
        )
        done()
      } catch (error) {
        done(error)
      }
    }, resizeHandleTimeout)
  })

  test("Adaptive should select minimal coefficient if two bases passed", done => {
    const width = 250
    resizeViewport(width, 299)
    setTimeout(() => {
      try {
        const rem = (width / 200) * 10
        expect(localVue.adaptive.rem).toBeCloseTo(rem)
        done()
      } catch (error) {
        done(error)
      }
    }, resizeHandleTimeout)
  })
})


jest.config.js
/**************************************************************************************************
 * Copyright © 2017-2021 Mikhail Perelygin <mk.prlgn@gmail.com>.                                  *
 * Huge thanks to:                                                                                *
 *     Sergey Koshevarov <gondragos@gmail.com> for plugin idea;                                   *
 *     Maxim Logvinov <skovorodker.rekdo@gmail.com> for optimisation ideas;                       *
 * Program is distributed under the terms of the MIT License.                                     *
 *                                                                                                *
 * @date        2.2.2021                                                                          *
 * @license     MIT                                                                               *
 **************************************************************************************************/

module.exports = {
  setupFilesAfterEnv: ["jest-extended"],
  moduleFileExtensions: ["js", "json", "vue"],
  transform: {
    ".*\\.(vue)$": "<rootDir>/node_modules/vue-jest",
    ".+\\.(js)$": "<rootDir>/node_modules/babel-jest"
  },
  collectCoverage: true,
  collectCoverageFrom: ["**/src/**/*.js", "!**/node_modules/**"]
}


src/adaptive.js
/*************************************************************************************************
 * Copyright © 2017-2018 Mikhail Perelygin <zmey3301@gmail.com>. All rights reserved.
 * Huge thanks to:
 *     Sergey Koshevarov <gondragos@gmail.com> for plugin idea;
 *     Maxim Logvinov <skovorodker.rekdo@gmail.com> for optimisation ideas;
 * Program is distributed under the terms of the GNU Affero General Public License.
 *
 *  @date         10.5.2018
 *  @license     AGPL-3.0-or-later
 ************************************************************************************************/
import defaults from 'lodash.defaults'
import throttle from 'lodash.throttle'

export default class {
	/**
	 * Splitting configuration and creating eventListeners
	 * @param Vue
	 * @param config: adaptive configuration file
	 * @returns {Record<string, any>}: reactive model, based on Vue
	 */
	constructor (Vue, config) {
		// Defaults
		const defaultGlobal = {
			throttle: 17,
			orientationTestCount: 50,
			orientationChangeTimeout: 1000
		}
		const defaultConfig = {
			if: false,
			element: window,
			rem: false,
			from: false,
			to: false,
			base: false,
			k: 1
		}
		// Global config
		this.globals = defaults(config.global, defaultGlobal)
		delete config.global
		let deviceList = {}
		for (let device in config) {
			if (config.hasOwnProperty(device)) {
				defaults(config[device], defaultConfig)
				device = device.split(':')[0]
				if (!deviceList.hasOwnProperty(device)) deviceList[device] = false
			}
		}
		this.config = config
		let data = {
			is: deviceList,
			width: 0,
			height: 0,
			rem: 0
		}
		this.VM = new Vue({ data })

		// Listening to change events
		window.addEventListener('resize', throttle(this.resize.bind(this, false), this.globals.throttle))
		window.addEventListener('orientationchange', this.orientationChange.bind(this))
		// Initializing Adaptive
		this.resize(true)

		return this.VM.$data
	}

	/**
	 * Handle resize of viewport
	 * @param init: initialize flag
	 */
	resize (init = false) {
		const html = document.documentElement
		let cache = {
			window: {
				width: html.offsetWidth,
				height: html.offsetHeight
			}
		}
		let newDeviceList = {}
		let rem = 10
		// Setting viewport size
		if (this.VM.$data.width !== cache.window.width) this.VM.$data.width = cache.window.width
		if (this.VM.$data.height !== cache.window.height) this.VM.$data.height = cache.window.height
		for (let name in this.config) {
			if (this.config.hasOwnProperty(name)) {
				let device = this.config[name]
				// Getting device name
				name = name.split(':')[0]
				// Caching elements viewport
				let data
				if (typeof device.element === 'string' && !cache.hasOwnProperty(device.element)) {
					let el = document.querySelector(device.element)
					data = cache[device.element] = {
						width: el.innerWidth,
						height: el.innerHeight
					}
				} else if (typeof device.element === 'string') data = cache[device.element]
				else if (typeof device.element !== 'string') data = cache['window']
				// Detecting is breakpoints valid
				// Testing if function
				let checked = !(device.if && typeof device.if === 'function' && !device.if(data) ||
					// Testing min viewport
					device.from && (device.from.width > data.width || device.from.height > data.height) ||
					// Testing max viewport
					device.to && (device.to.width <= data.width || device.to.height <= data.height))
				// Testing classes
				if (!newDeviceList.hasOwnProperty(name) || !newDeviceList[name]) newDeviceList[name] = checked
				// Scale changing
				if (checked) {
					// Setting static rem
					if (device.rem) {
						rem = device.rem
						// Setting dynamic rem
					} else if (device.base) {
						let remBases = []
						if (device.base.width) {
							remBases.push(data.width / device.base.width)
						}
						if (device.base.height) {
							remBases.push(data.height / device.base.height)
						}
						let remBase = Math.min(...remBases)
						let k = device.k || 1
						rem = remBase * k * 10
					}
				}
			}
		}
		// Setting device
		for (let name in newDeviceList) {
			if (newDeviceList.hasOwnProperty(name)) {
				let checked = newDeviceList[name]
				let oldClass = this.VM.$data.is[name] ? name : `no-${name}`
				let newClass = checked ? name : `no-${name}`
				// Updating classes if changed
				if (oldClass !== newClass || init) {
					if (!init) html.classList.remove(oldClass)
					html.classList.add(newClass)
					this.VM.$data.is[name] = checked
				}
			}
		}
		if (rem !== this.VM.$data.rem) {
			html.style.fontSize = `${rem}px`
			this.VM.$data.rem = rem
		}
	}

	/**
	 * Handle orientationChange event
	 */
	orientationChange () {
		let noChangeCount = 0

		let end = function () {
			clearInterval(interval)
			clearTimeout(timeout)
			interval = null
			timeout = null
		}
		let html = document.documentElement
		let interval = setInterval(() => {
			let currHeight = html.offsetHeight
			let currWidth = html.offsetWidth
			if (currWidth === this.VM.$data.width && currHeight === this.VM.$data.height) {
				noChangeCount++
				if (noChangeCount === this.globals.orientationTestCount) {
					end()
				}
			} else {
				this.VM.$data.width = currWidth
				this.VM.$data.height = currHeight
				this.resize()
				noChangeCount = 0
			}
		}, this.globals.throttle)
		let timeout = setTimeout(function () {
			end()
		}, this.globals.orientationChangeTimeout)
	}
}


src/index.js
/**************************************************************************************************
 * Copyright © 2017-2021 Mikhail Perelygin <mk.prlgn@gmail.com>.                                  *
 * Huge thanks to:                                                                                *
 *     Sergey Koshevarov <gondragos@gmail.com> for plugin idea;                                   *
 *     Maxim Logvinov <skovorodker.rekdo@gmail.com> for optimisation ideas;                       *
 * Program is distributed under the terms of the MIT License.                                     *
 *                                                                                                *
 * @date        28.1.2021                                                                         *
 * @license     MIT                                                                               *
 **************************************************************************************************/

import ClassList from "classlist"
import throttle from "lodash.throttle"

export default class Adaptive {
  /**
   * Viewport describing object
   * @typedef viewport
   * @type {object} - viewport size
   * @property {number} width - viewport width
   * @property {number} height - viewport height
   */

  /**
   * Device configuration object
   * @typedef device
   * @type {object}
   * @property {function({width: number, height: number})} [if] - custom device condition callback
   * @property {string|HTMLElement} [element=window] - element selector that will be used for detecting viewport
   * @property {number} [rem] - static rem value, overrides k and base calculation
   * @property {number} [k=1] - Additional coefficient for rem calculation
   * @property {viewport} [base] - base width and height for rem calculation
   * @property {viewport} [from] - minimal width and height for setting device
   * @property {viewport} [to] - maximal width and height for setting device
   * @property {boolean} [setDevice=true] - class, no-class and device set toggle
   */

  /**
   * Plugin configuration object
   * @typedef globalConf
   * @type {object}
   * @property {number} [rem=10] - default rem value in case no breakpoints matched
   * @property {number} [throttle=17] - frequency of viewport data update
   * @property {number} [orientationTestCount=50] - orientationchange event end detection tries
   * @property {number} [orientationChangeTimeout=1000] - maximum orientationchange event duration before viewport update
   */

  /**
   * Current adaptive data
   * @typedef adaptiveObservable
   * @type {object}
   * @property {object<string,boolean>} is - devices status
   * @property {number} width - current viewport width
   * @property {number} height - current viewport height
   * @property {number} rem - current rem size in px
   */

  /**
   * Vue plugin install function
   * @param Vue - Vue framework object
   * @param {object<string, device>} config - adaptive configuration file
   * @param {globalConf} config.global - Plugin's instance-level config
   */
  static install(Vue, config) {
    const adaptive = new this(Vue, config)
    Vue.adaptive = adaptive
    Object.defineProperty(Vue.prototype, "$adaptive", {
      /**
       * Adaptive current state getter
       * @returns {Adaptive} - Adaptive current state observable
       */
      get: () => adaptive
    })
  }

  /**
   * Splitting configuration and creating eventListeners
   * @param Vue - Vue framework object
   * @param {{string: device}} config - adaptive configuration file
   * @param {globalConf} config.global - Plugin's instance-level config
   * @returns {adaptiveObservable} - reactive model, based on Vue
   */
  constructor(Vue, config) {
    // Defaults
    const defaultGlobal = {
      rem: 10,
      throttle: 17,
      orientationTestCount: 25,
      orientationChangeTimeout: 1000
    }
    const defaultConfig = {
      if: null,
      element: window,
      rem: null,
      from: null,
      to: null,
      base: null,
      k: 1,
      setDevice: true
    }
    // Global config
    this.globals = Object.assign(defaultGlobal, config.global)
    this.config = {}

    const deviceList = {}
    for (let device in config) {
      if (config.hasOwnProperty(device) && device !== "global") {
        const deviceConfig = (this.config[device] = Object.assign(
          {},
          defaultConfig,
          config[device]
        ))

        device = device.split(":")[0]
        if (deviceConfig.setDevice && !deviceList.hasOwnProperty(device))
          deviceList[device] = false
      }
    }
    const data = {
      is: deviceList,
      width: 0,
      height: 0,
      rem: 0
    }
    const [major, minor] = Vue.version.split(".").slice(0, 2).map(Number)
    /** @type adaptiveObservable */
    this.data =
      major === 2 && minor >= 6 ? Vue.observable(data) : new Vue({data}).$data

    // Listening to change events
    window.addEventListener(
      "resize",
      throttle(this.resize.bind(this), this.globals.throttle)
    )
    window.addEventListener(
      "orientationchange",
      this.orientationChange.bind(this)
    )
    // Initializing Adaptive
    this.resize()

    return this.data
  }

  /**
   * Handle resize of viewport
   */
  resize() {
    const viewport = {
      width: window.innerWidth,
      height: window.innerHeight
    }
    const cache = {
      keys: [window],
      values: [viewport]
    }
    const newDeviceList = {}
    let rem = this.globals.rem
    // Setting viewport size
    if (this.data.width !== viewport.width) this.data.width = viewport.width
    if (this.data.height !== viewport.height) this.data.height = viewport.height

    for (let name in this.config) {
      if (this.config.hasOwnProperty(name)) {
        const device = this.config[name]
        // Getting device name
        name = name.split(":")[0]
        // Caching elements viewport
        const elementCacheIndex = cache.keys.indexOf(device.element)
        let data
        if (elementCacheIndex === -1) {
          const el =
            device.element instanceof HTMLElement
              ? device.element
              : document.querySelector(device.element)
          const clientRect = el.getBoundingClientRect()
          data = {
            width: clientRect.width,
            height: clientRect.height
          }
          cache.keys.push(device.element)
          cache.values.push(data)
        } else data = cache.values[elementCacheIndex]
        // Detecting is breakpoints valid
        // Testing if function
        const checked = !(
          (typeof device.if === "function" && !device.if(data)) ||
          // Testing min viewport
          (device.from &&
            (device.from.width > data.width ||
              device.from.height > data.height)) ||
          // Testing max viewport
          (device.to &&
            (device.to.width <= data.width || device.to.height <= data.height))
        )
        // Testing classes
        if (device.setDevice && !newDeviceList[name])
          newDeviceList[name] = checked
        // Scale changing
        if (checked) {
          // Setting static rem
          if (device.rem) {
            rem = device.rem
            // Setting dynamic rem
          } else if (device.base) {
            const remBases = []
            if (device.base.width) {
              remBases.push(data.width / device.base.width)
            }
            if (device.base.height) {
              remBases.push(data.height / device.base.height)
            }
            const remBase = Math.min(...remBases)
            rem = remBase * device.k * 10
          }
        }
      }
    }

    this.updateDOM(newDeviceList, rem)
  }

  /**
   * Update documentElement classes and font size
   * @param {object.<string,boolean>} newDeviceList - new devices state
   * @param {number} rem - new rem size calculation
   */
  updateDOM(newDeviceList, rem) {
    const documentElementClassList = new ClassList(document.documentElement)

    for (const name in newDeviceList) {
      if (newDeviceList.hasOwnProperty(name)) {
        const checked = newDeviceList[name]
        const oldClass = this.data.is[name] ? name : `no-${name}`
        const newClass = checked ? name : `no-${name}`
        // Updating classes if changed
        if (
          oldClass !== newClass ||
          !documentElementClassList.contains(newClass)
        ) {
          if (documentElementClassList.contains(oldClass))
            documentElementClassList.remove(oldClass)
          documentElementClassList.add(newClass)
          this.data.is[name] = checked
        }
      }
    }
    if (rem !== this.data.rem) {
      document.documentElement.style.fontSize = `${rem}px`
      this.data.rem = rem
    }
  }

  /**
   * Handle orientationChange event
   */
  orientationChange() {
    let noChangeCount = 0

    const end = function () {
      clearInterval(interval)
      clearTimeout(timeout)
      interval = null
      timeout = null
    }
    let interval = setInterval(() => {
      const currHeight = window.innerHeight
      const currWidth = window.innerWidth
      if (currWidth === this.data.width && currHeight === this.data.height) {
        noChangeCount++
        if (noChangeCount === this.globals.orientationTestCount) end()
      } else {
        this.data.width = currWidth
        this.data.height = currHeight
        this.resize()
        noChangeCount = 0
      }
    }, this.globals.throttle)
    let timeout = setTimeout(function () {
      end()
    }, this.globals.orientationChangeTimeout)
  }
}
